src-plugins/mscAlgorithmPaint/mscAlgorithmPaintToolBox.cpp

```diff
@@ -1445,48 +1456,29 @@ void AlgorithmPaintToolBox::addSliceToStack(medAbstractView * view,const unsigne
             m_redoStacks->value(view)->clear();
 }
 
-void AlgorithmPaintToolBox::closeView()
+void AlgorithmPaintToolBox::clearToolBox()
 {
-    // close every data in the view
-    medAbstractView* view = qobject_cast<medAbstractView*>(QObject::sender());
-    medAbstractLayeredView* layeredView = dynamic_cast<medAbstractLayeredView*>(view);
-
-    if (layeredView->layersCount() > 0)
+    if (currentView && (currentView->layersCount()>0))
     {
-        // remove a data, which will recursively call this closeView() function
-        // until there is no more data in the view
-        layeredView->removeLayer(0);
-    }
-    else
-    {
-        // clean the toolbox when there is no data left in the view
-        onViewClosed(view);
-    }
-}
-
-void AlgorithmPaintToolBox::onViewClosed(medAbstractView* viewClosed)
-{
-    clearMask();
-
-    if (m_undoStacks->value(viewClosed))
-    {
-        m_undoStacks->value(viewClosed)->clear();
-        m_redoStacks->value(viewClosed)->clear();
-        m_undoStacks->remove(viewClosed);
-        m_redoStacks->remove(viewClosed);
+        // remove each "no name" mask data in the view
+        for(int unsigned i=0; i<currentView->layersCount(); i++)
```

Comment:
The end condition of this loop is not fixed. ```layersCount()``` will  decrease each time you remove a layer, and so you will miss some.